[HUDI-836] Implement datadog metrics reporter

xushiyan opened the pull request:
* Add a new metric reporter type for Datadog
* Implement reporting support for Datadog http api

## Committer checklist

 - [x] Has a corresponding JIRA in PR title & commit
 
 - [x] Commit message is descriptive of the change
 
 - [ ] CI is green

 - [ ] Necessary doc changes done or have another open PR
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

xushiyan commented:
@yanghua This is ready for review. Also manually tested the implementation and was able to collect metrics in Datadog.

![Screen Shot 2020-05-05 at 5 03 14 PM](https://user-images.githubusercontent.com/2701446/81127259-8a323f00-8ef2-11ea-9e92-b5350e2d64aa.png)

xushiyan commented:
@vinothchandar I'm thinking of writing another short blog to introduce Datadog support with the new configs. Sounds good?

yanghua commented:
> @vinothchandar I'm thinking of writing another short blog to introduce Datadog support with the new configs. Sounds good?

@xushiyan Absolutely good idea, IMO, not only blog but also documentation of config option.

xushiyan commented:
@yanghua Right. I'll make another PR to update the configs docs, too.

xushiyan commented:
Just observed weird CI failure...first time seeing OOM there. Trying with some settings in
https://github.com/apache/incubator-hudi/pull/1595

vinothchandar commented:
@xushiyan absolutely... please go ahead!

xushiyan commented:
@yanghua @vinothchandar The travis tests is flaky; i can pass the tests locally. Please go ahead review the PR; I'll do some checkings on the CI failure later. Thank you.

vinothchandar commented:
@xushiyan I think the metrics tests are hard coding the ports and running into issues? I ran into one such issue locally.. PR brewing..

xushiyan commented:
@vinothchandar @yanghua CI is passing now... I don't seem to have tests failing; it seemed just sometimes travis got freaked out...

xushiyan commented:
@yanghua Once this is approved, I'll then update this docs update PR https://github.com/apache/incubator-hudi/pull/1603 with Chinese version. Thanks.

yanghua commented:
> @yanghua Once this is approved, I'll then update this docs update PR #1603 with Chinese version. Thanks.

OK, will review soon.

yanghua commented on hudi-client/src/main/java/org/apache/hudi/config/HoodieMetricsConfig.java:
```diff
@@ -48,6 +48,8 @@
   public static final String GRAPHITE_SERVER_PORT = GRAPHITE_PREFIX + ".port";
   public static final int DEFAULT_GRAPHITE_SERVER_PORT = 4756;
 
```
remove this empty line?

yanghua commented on hudi-client/src/main/java/org/apache/hudi/config/HoodieMetricsConfig.java:
```diff
@@ -92,28 +103,63 @@ public Builder withReporterType(String reporterType) {
     }
 
-    public Builder toGraphiteHost(String host) {
+    public Builder withGraphiteHost(String host) {
```
I have two opinions about these changes:

- Does it break the compatibility?
- Packaging different config options of different reporters is not a good choice, it would be better to refactor.

xushiyan replied:
These builder methods are not used by any logic. I think they are legacy code and only mentioned in config docs. I'm changing the names in #1603 too. 

You're right that we should segregate the configs for different reporter types. It shouldn't give too much hassle to do refactoring.

yanghua commented on hudi-client/src/main/java/org/apache/hudi/config/HoodieWriteConfig.java:
```diff
@@ -507,6 +512,37 @@ public String getJmxPort() {
     return props.getProperty(HoodieMetricsConfig.JMX_PORT);
   }
 
+  public ApiSite getDatadogApiSite() {
```
Ditto, IMO, we should also refactor these change in the future.

yanghua commented on hudi-client/src/main/java/org/apache/hudi/metrics/datadog/DatadogHttpClient.java:
```diff
@@ -0,0 +1,128 @@
+  private static final String SERIES_URL_FORMAT = "https://app.datadoghq.%s/api/v1/series";
+  private static final String VALIDATE_URL_FORMAT = "https://app.datadoghq.%s/api/v1/validate";
+  private static final String HEADER_KEY_API_KEY = "DD-API-KEY";
+  private static final int TIMEOUT_MILLIS = 3000;
```
Can we make this field configurable?

yanghua commented on hudi-client/src/main/java/org/apache/hudi/metrics/datadog/DatadogMetricsReporter.java:
```diff
@@ -0,0 +1,90 @@
+
+  @Override
+  public void start() {
+    reporter.start(30, TimeUnit.SECONDS);
```
Can this hard code be configurable?

xushiyan replied:
I tried to gauge on a good number of configs for user to control over.. but i guess it won't harm if we go with all configurable with sane defaults. So yes i can make it configurable.

yanghua commented on hudi-client/src/test/java/org/apache/hudi/metrics/datadog/TestDatadogHttpClient.java:
```diff
@@ -0,0 +1,152 @@
+  }
+
+  @Test
+  public void validateApiKey_shouldThrowException_whenRequestFailed() throws IOException {
```
It may be my lack of knowledge. Are these patterns like `xxx_xxx_xxx` related to the parsing mechanism of the latest test framework? Otherwise, why not use hump nomenclature?

xushiyan replied:
It is just one unit test naming convention I tried to adopt...it basically represents [theMethodUnderTest_expectedBehavior_condition] for being descriptive of the logic. See [this](https://stackoverflow.com/questions/155436/unit-test-naming-best-practices) for reference.

The `_` is to break the long sentence for readability when using hump nomenclature. I don't mind changing it to hump nomenclature though.

yanghua replied:
I guess it may be a practice. I do not against this style, it looks good. But for now, it would be better to keep a unified style right?

Of cause, you can start a DISCUSS  on dev ML.

xushiyan replied:
Yup changed to hump namings.

yanghua commented on pom.xml:
```diff
@@ -245,6 +245,7 @@
         <version>${maven-surefire-plugin.version}</version>
         <configuration>
           <skip>${skipUTs}</skip>
+          <argLine>-Xms256m -Xmx2g</argLine>
```
We must add this line in this PR?

xushiyan replied:
it was there when i got OOM in travis...maybe not needed now. let me try removing it

xushiyan commented:
@yanghua In the last commit, I added a new config class `HoodieMetricsDatadogConfig` just for datadog related configs and reverted previous change to `HoodieMetricsConfig`. I think it'd be better to refactor the configs of 2 other reporter types in a separate PR to minimize risks. If the new config class looks good, I'll do a separate PR for the other two, probably with some test cases, too.

xushiyan commented:
@yanghua I kept getting OOM; have to set surefire with more heap.. do you prefer me do another MINOR PR to increase the heap?

yanghua commented:
> @yanghua I kept getting OOM; have to set surefire with more heap.. do you prefer me do another MINOR PR to increase the heap?

Yes, It would be better to change it on a single PR.

xushiyan commented:
> > @yanghua I kept getting OOM; have to set surefire with more heap.. do you prefer me do another MINOR PR to increase the heap?
> 
> Yes, It would be better to change it on a single PR.

OK created https://github.com/apache/incubator-hudi/pull/1623

Does the latest change look good? If yes, I can refactor the configs as it's done in this PR. Thanks.

yanghua commented:
@xushiyan I have no other concerns.

xushiyan commented:
Note: blocked by #1623

xushiyan commented:
@yanghua CI passed. Also updated #1603 for the newly added config. Thanks.

yanghua commented:
Will do a final check.

xushiyan commented on hudi-client/src/main/java/org/apache/hudi/metrics/datadog/DatadogReporter.java:
```diff
@@ -0,0 +1,168 @@
+    }
+
+    PayloadBuilder addGauge(String metric, long timestamp, long gaugeValue) {
+      ValidationUtils.checkState(Objects.equals(type, "gauge"));
```
@yanghua fixed.

yanghua commented:
@vinothchandar If you are busy with other things. Can we merge it firstly? Then, iterate later.

vinothchandar commented:
Sorry folks.. Got delayed ..  This looks fine at a high level.. I was mulling if using http-client directly will cause any issues for us.. but seems ok.. 

If something comes up, we can fix forward..

vinothchandar commented:
@xushiyan may be we need to doc this or write a small post?

xushiyan commented:
> @xushiyan may be we need to doc this or write a small post?

Yes @vinothchandar will do the post soon. Config docs update is in #1603 . As for http-client, I have been using this implementation in production, looking good so far. Will definitely test it with more scenarios.
